---
title: "ADMIXTURE Workflow"
author: "Katherine Drotos"
date: "2024-10-23"
output: md_document:
  variant: markdown_github
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval=FALSE)
```

## Introduction

This is a workflow for the use of [ADMIXTURE](https://dalexander.github.io/admixture/index.html) (Alexander et al. 2009) in a Linux environment, geared towards novice users. It assumes some knowledge of bash.

ADMIXTURE is well used and there are several good resources to help you:

-   [ADMIXTURE 'official' documentation](https://dalexander.github.io/admixture/admixture-manual.pdf)

-   [Joana Meier's excellent how-to guide](https://speciationgenomics.github.io/ADMIXTURE/)

## 1.0 Environment Setup

We work on a compute cluster using a Slurm scheduler, so some setup here will be specific to that. Our cluster also already had ADMIXTURE installed.

If you are planning on using ADMIXTURE on your local machine, you can [download them via the binaries](https://dalexander.github.io/admixture/download.html) on the ADMIXTURE website. Then you will either need to work in the directory where ADMIXTURE is installed, or provide the full pathway to call the software.

For our setup, a few modules on the cluster are required to be loaded in, followed by ADMIXTURE itself:

```{bash}
module load StdEnv/2020
module load nixpkgs/16.09

module load admixture/1.3.0
```

We did not experiment much with different module versions here. ADMIXTURE is reasonably flexible with different environments (unlike, say, fastSTRUCTURE) so hopefully whatever your default setup is will work. If not, you can search for the setup modules that are missing from your machine (an ADMIXTURE error may tell you what you need), and potentially run everything using a virtual environment.

## 2.0 Using ADMIXTURE

### 2.1 Preparing input files

Ensure your input data files are in the same folder/directory as ADMIXTURE, or call the full path. ADMIXTURE accepts .bed format . For example, assuming you have a .vcf, you can use `plink` (which you will need to install if you don't already have it) to convert to .bed:

```{bash}
plink --vcf data_file.vcf --make-bed --out data_file --allow-extra-chr
```

This gives `data_file.bed` as a converted file ready for input into ADMIXTURE. Not that the associated .bim and .fam files will also need to be in the same directroy for ADMIXTURE to read the .bed file correctly (the `plink` command above automatically produces all three files at once).

### 2.2 Running ADMIXTURE

The basic ADMIXTURE run command looks like this:

```{bash}
admixture data_file.bed $k
```

where `$k` corresponds to the number of clusters being tested in the model.

There are several other parameters you may want to use:

| Parameter | What it does |
|----------------|--------------------------------------------------------|
| `-B` | Bootstraps the standard errors; default is 200, and more can be specified by including the number immediately after the B (e.g., `-B2000`). NOTE: this requires a .map file to work. |
| `--cv` | Cross validation error; this is one method to determine *k*. Ideally, the most optimal *k* will have the lowest cv error. This defaults to 5-fold, but you can increase it (e.g., `â€“cv=10`). |
| `-s` | Changes the random seed. Note that by default, ADMIXTURE uses the same random seed every time, so you will want to change this if doing multiple reps (e.g., `-s 30450`) |

For a complete example, let's say you had an input file and a .map file, and wanted to test *k*=3, with standard error bootstrapping of 1,000 and a cross validation of 10-fold. With a random seed, that would look like:

```{bash}
admixture -B1000 --cv=10 -s 46921 data_file.bed 3
```

ADMIXTURE generates five output files: the log file, .P, .Q, .Q_bias, and .Q_se. To my knowledge, you can only specify the name of the log file, like so:

```{bash}
admixture -B1000 --cv=10 -s 46921 data_file.bed 3 > ADMIXTUREanalysis_mydata_k3.out
```

All the other output files will be named for the input file, as: data_file.3.Q, where the number is the value of *k* tested.

This is important to remember when you are running multiple replicates in the same directory. Since all output files (other than the log file) will have a default name using the input file and your tested *k* value, if you run multiple reps for a given *k* then they will overwrite each other. Therefore you will want to either move output files to another directory before starting the next run, or start the runs from separate directories in the first place.

### 2.3 Looping ADMIXTURE over values of *k*

To start runs across multiple values of *k* from one command, you can loop it like so:

```{bash}
for k in {1..5}; do
admixture -B1000 --cv=10 -s 46921 data_file.bed $k > ADMIXTUREanalysis_mydata_${k}.out
```

Let's say you wanted 10 replicates per tested value of *k*. You could run this command ten times (either moving the output files each time, or running from different directories) with just changing the `-s` seed and the log file name to identify the replicate.

## 3.0 Estimating *k*

### 3.1 Original method (Alexander et al. 2009)

Per the ADMIXTURE documentation, the original method to determine which *k* value is optimal for your data is to look for the lowest cross validation error across runs. The cv error is found in the .out log file. From a given directory, you can copy all cv errors to a new file like so:

```{bash}
grep -i "CV error" ADMIXTUREanalysis_mydata_* > ADMIXTURErun01_cv
```

You can then concatenate all your cv errors across all your replicate runs for further analysis, by:

```{bash}
cat ADMIXTURErun* >> ADMIXTURE_allruns_cv.txt
```

### 3.2 Evanno method (repurposing Evanno et al. 2005)

From our experience, we found that using cross validation error alone was not reliable, and so began using the Evanno method (originally designed for STRUCTURE) to determine the optimal *k* value. This method is covered in the [ADMIXTURE_Evanno_chooseK_workflow](https://github.com/kdrotos/Cluster_tools_comparison/blob/main/ADMIXTURE_Evanno_chooseK_workflow.Rmd), and testing and justification is in our forthcoming publication.
